MetaMask/metamask-extension
# fix: Fix issue 25285 max insufficient funds for gas

## **Description**

When performing a send or swap+send there are two outstanding issues:
1. swap+send, nativeToken -> ERC20 and "insufficient funds for gas" error message appears. This behavior is incorrect estimation and should only occur when there are insufficient funds for simple send. In other words, the current logic is treating nativeToken -> ERC20 swaps as if they were nativeToken -> nativeToken swaps, which is causing this error.
1. when swapping between src assets the amount mode max is maintained. Amounts should reset between differently chosen srcAssets.

Fixes are:
* Hiding the max button when the source token is native, but the destination token is an ERC20
* Resetting the srcAmount whenever the srcAsset changes




[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25574?quickstart=1)

## **Related issues**

Fixes: 
* https://github.com/MetaMask/metamask-extension/issues/25285
* https://github.com/MetaMask/metamask-extension/issues/25594

## **Manual testing steps**

For each every combination of srcToken to destToken either being nativeToken or an ERC20
Verify that the insufficient funds message does not show unless it is nativeToken -> nativeToken and there is insufficient funds for gas

when maxing srcToken (current behavior)
| srcToken | destToken | 
| --- | --- | 
| nativeToken | nativeToken |
| nativeToken | ERC20 |  
| ERC20 | nativeToken | 
| ERC20 | ERC20 |

1.  click on 'Send'
3. perform a 'Send' or 'Swap + Send' based on the above table
4. verify that the error message appears only in the situation where there is insufficient native token to pay for gas for a 'Send'




## **Screenshots/Recordings**



### **Before**

<img width="1187" alt="Screenshot 2024-06-20 at 00 01 00" src="https://github.com/MetaMask/metamask-extension/assets/104780023/1d5dd5f7-3ad2-4f8b-a9a5-503927070458">


### **After**

<img width="200" alt="Screenshot 2024-06-28 at 15 15 43" src="https://github.com/MetaMask/metamask-extension/assets/3590816/d5645dba-dc91-4da7-b909-bec2440131ca">
<img width="200" alt="Screenshot 2024-06-28 at 15 15 46" src="https://github.com/MetaMask/metamask-extension/assets/3590816/d66f0906-581b-48cd-b1b2-08f78a961670">
<img width="200" alt="Screenshot 2024-06-28 at 15 15 53" src="https://github.com/MetaMask/metamask-extension/assets/3590816/0bb74da7-41d6-4004-acbe-cb263fbef411">
<img width="200" alt="Screenshot 2024-06-28 at 15 15 56" src="https://github.com/MetaMask/metamask-extension/assets/3590816/e3e20655-bb05-4ad6-b90f-bda96d42c677">
<img width="200" alt="Screenshot 2024-06-28 at 15 16 16" src="https://github.com/MetaMask/metamask-extension/assets/3590816/75e766fe-a1a0-4709-ad6f-1683762c64f2">
<img width="200" alt="Screenshot 2024-06-28 at 15 16 24" src="https://github.com/MetaMask/metamask-extension/assets/3590816/fdfab83f-4d69-45da-8ca7-81ae56c9538e">
<img width="200" alt="Screenshot 2024-06-28 at 15 16 32" src="https://github.com/MetaMask/metamask-extension/assets/3590816/39b4315d-cd86-4f63-9edb-8ed838761e72">

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
